src/middleware/graphql.ts

```diff
@@ -55,38 +60,70 @@ export abstract class GraphQLBase implements Initializable<express.Application>
     this.remoteSchemas = this.createRemoteSchemaMap(config.remoteSchemaUrls);
     this.createDataLoaders = config.createDataLoaders;
     this.schemaUrls = config.remoteSchemaUrls;
+    this.retryOperations = {};
+    this.remoteSchemaVersions = {};
+  }
+
+  get hasRemoteSchema() {
+    return Boolean(this.schemaUrls) && Boolean(this.schemaUrls.length);
   }
 
   public async initialize(app: express.Application) {
-    await this.resolveRemoteSchemas();
+    await this.resolveRemoteSchemas(this.remoteSchemas);
     this.refreshMiddleware();
     this.graphqlMiddleware.handler();
     this.server.applyMiddleware({ app, path: this.path });
     return Promise.resolve();
   }
 
   public isSchemaHealthy() {
-    return !this.schemaUrls || this.schemaUrls.every(url => !!this.remoteSchemas[url]);
+    return !this.hasRemoteSchema || this.schemaUrls.every(url => Boolean(this.remoteSchemas[url]));
   }
 
-  public async updateRemoteSchema(url: string) {
+  public async updateRemoteSchema(url: string, targetVersion?: string) {
     log.info(`Updating remote schema at ${url}`);
-    await this.addRemoteSchema(url);
-    this.refreshMiddleware();
+
+    // When passing a schema version if the version matches no need to update
+    if (this.isVersionCorrect(url, targetVersion)) {
+      return Promise.resolve();
+    }
+
+    this.sendFetchingEvent();
+
+    if (await this.addRemoteSchema(url, targetVersion)) {
+      this.refreshMiddleware();
+    }
+
+    this.shouldSendFetchingCompleted();
+    return Promise.resolve();
   }
 
   public getRemoteSchemaStatus(url: string) {
     return this.remoteSchemas[url] ? RemoteSchemaStatus.ACTIVE : RemoteSchemaStatus.INACTIVE;
   }
 
-  private async resolveRemoteSchemas() {
-    await Promise.all(Object.keys(this.remoteSchemas).map(url => this.addRemoteSchema(url)));
+  private isVersionCorrect(url: string, targetVersion: string) {
+    return Boolean(targetVersion) && this.remoteSchemaVersions[url] === targetVersion;
+  }
+
+  private async resolveRemoteSchemas(remoteSchemas: GraphQLSchemaMap) {
+    this.sendFetchingEvent();
 
+    await Promise.all(Object.keys(remoteSchemas).map(url => this.addRemoteSchema(url)));
+
+    this.shouldSendFetchingCompleted();
+
+    return Promise.resolve();
+  }
+
+  private sendFetchingEvent() {
+    this.publish(ServerLifeCycle.REMOTE_SCHEMAS_FETCHING);
+  }
+
+  private shouldSendFetchingCompleted() {
```

Comment:
This method name implies a check that returns a boolean result, rather than an action being taken. At the same time, since the action won't always happen, the name shouldn't imply that it will definitely happen. A name that accurately conveys what's happening here might be `maybeSendFetchingCompleted`, though the `maybe` part doesn't really say what that depends on. What do you think?